numpyro/distributions/discrete.py

```diff
@@ -1,171 +1,393 @@
+            raise ValueError("`logits` parameter must be at least one-dimensional.")
+        batch_shape = lax.broadcast_shapes(np.shape(logits)[:-1], np.shape(total_count))
+        logits = logits - logsumexp(logits)
+        self.logits = promote_shapes(logits, shape=batch_shape + np.shape(logits)[-1:])[0]
+        self.total_count = promote_shapes(total_count, shape=batch_shape)[0]
+        super(MultinomialWithLogits, self).__init__(batch_shape=batch_shape,
+                                                    event_shape=np.shape(self.logits)[-1:],
+                                                    validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return multinomial_rvs(key, self.total_count, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        dtype = get_dtypes(self.probs)[0]
+        value = lax.convert_element_type(value, dtype)
+        total_count = lax.convert_element_type(self.total_count, dtype)
+        logits = self.logits
+        logits = np.clip(logits, a_min=np.finfo(get_dtypes(logits)[0]).min)
+        return gammaln(total_count + 1) + np.sum(value * logits - gammaln(value + 1), axis=-1)
+
+    @lazy_property
+    def probs(self):
+        return _to_probs_multinom(self.logits)
+
+    @property
+    def mean(self):
+        return np.broadcast_to(np.expand_dims(self.total_count, -1) * self.probs,
+                               self.batch_shape + self.event_shape)
+
+    @property
+    def variance(self):
+        return np.broadcast_to(np.expand_dims(self.total_count, -1) * self.probs * (1 - self.probs),
+                               self.batch_shape + self.event_shape)
+
+    @property
+    def support(self):
+        return constraints.multinomial(self.total_count)
+
+
+class Categorical(Distribution):
+    arg_constraints = {'probs': constraints.simplex}
+
+    def __init__(self, probs, validate_args=None):
+        if np.ndim(probs) < 1:
+            raise ValueError("`probs` parameter must be at least one-dimensional.")
+        self.probs = probs
+        super(Categorical, self).__init__(batch_shape=np.shape(self.probs)[:-1],
+                                          validate_args=validate_args)
+
+    def sample(self, key, size=()):
+        return categorical_rvs(key, self.probs, shape=size + self.batch_shape)
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        batch_shape = lax.broadcast_shapes(np.shape(value), self.batch_shape)
+        value = np.expand_dims(value, axis=-1)
+        value = np.broadcast_to(value, batch_shape + (1,))
+        log_pmf = np.broadcast_to(self.logits, batch_shape + np.shape(self.logits)[-1:])
+        return np.take_along_axis(log_pmf, value, axis=-1)[..., 0]
+
+    @lazy_property
+    def logits(self):
+        return _to_logits_multinom(self.probs)
```

Comment:
I don't see `logits` property in other discrete distribution. Could you add a comment for why it is needed in the computation of `log_prob`?

Reply:
`logits` is nothing but a clipped version of `np.log(probs)` for the multinomial case. The other distributions don't have it because they use the `xlogy` formulation in `log_prob`. That said, it might be a good idea to have this in other distributions too, since `log` is expensive and if this can be cached across multiple calls to `log_prob` we should do that.

Reply:
Hmm, I expect that there are many things to cache but currently we don't use that convention. For example, `gammaln(count)` is also expensive, `log(scale)` in normal/cauchy distributions are also expensive. Could you please remind me why we need to separate `Multinomial` and `MultinomialWithLogits`? (so I can decide what to do with MVN distribution)

Reply:
I agree with you - let us remove this and if we really need to cache computations, we should do it in a more principled way. 

> Could you please remind me why we need to separate Multinomial and MultinomialWithLogits? (so I can decide what to do with MVN distribution)

We don't need to (and we aren't in PyTorch) - it just seemed easier to not have to sprinkle the code with `if / else`, and have to put in optional parameters in `arg_contraints`. I think, within limits, if we have multiple parametrizations for a distribution, it is best to separate out the implementations. e.g. `OneHotCategorical` and `Categorical`, for instance. How is this relevant to MVN?

Reply:
In MVN, there are 3 args `scale_tril`, `covariance`, `precision` so there are 3 constraints to check. If we don't use `precision_matrix` then we should not check for it because it will be expensive taking inverse. I think that we just check for all of them as in PyTorch and ask users for providing suitable `validate_args` argument if they want to filter out some constraints.

Reply:
Hmm..the the only flip-side to having separate classes is code duplication. This is one of those cases where I think some code duplication might be okay (within limits, we don't want 10 MVN classes that end up confusing the user 😄), since it is not like we will be updating the distributions code every day. The advantage is that we don't have to figure about which property was provided and which property should not accessed (via `lazy_property`), etc. 

I will let you figure out how you want to go about it, either is fine with me, and I'm open to merging the logits/probs classes too.